client-apis/decrypt.md

```diff
@@ -69,65 +91,226 @@ This default CMM MUST obtain the decryption materials required for decryption.
+deserializing those bytes according to the [message format](../data-format/message.md).
+
+This operation MUST attempt to deserialize all available encrypted message bytes until it has
+successfully deserialized a valid [message header](../data-format/message-header.md).
+
+This operation MUST wait if it doesn't have enough encrypted message bytes available to
+deserialize the next field of the message header until enough input bytes become available or
+the caller indicates an end to the encrypted message.
 
-[TODO]
+Once a valid message header is deserialized,
+this operation MUST validate the [message header body](../data-format/message-header.md#header-body)
+by using the [authenticated encryption algorithm](../framework/algorithm-suites.md#encryption-algorithm)
+to decrypt with the following inputs:
 
-## Apendix
+- the AAD is the serialized [message header body](../data-format/message-header.md#header-body).
+- the IV has a value of 0
+- the cipherkey is the derived data key
+- the ciphertext is an empty byte array
+
+If this decryption fails, this operation MUST immediately halt and fail.
+
+If the algorithm suite contains a signature algorithm and
+this operation is streaming the plaintext output to the caller,
+this operation MUST input the serialized frame to the signature algorithm as soon as it is deserialized,
+such that the serialized frame isn't required to remain in memory to complete
+the [signature calculation](#signature-calculation).
+
+If this decryption succeeds and the algorithm suite does not contain a signature algorithm,
+the parsed [encryption context](#encryption-context) MAY be released to the caller.
+
+If the algorithm suite does contain a signature algorithm
+the parsed [encryption context](#encryption-context) MUST NOT be released to the caller until the
+[signature is verified](#verify-the-signature-optional).
+
+If streaming the output plaintext for a message with a signature,
+the result plaintext MAY be released, however the caller MUST NOT
+consider this plaintext verified.
+See [security considerations](#security-considerations) below.
+
+### Get the decryption materials
+
+To decrypt the message body, a set of valid decryption materials is required.
+
+This operation MUST obtain this set of [decryption materials](../framework/structures.md#decryption-materials),
+by calling [Decrypt Materials](../framework/cmm-interface.md#decrypt-materials) on a [CMM](../framework/cmm-interface.md).
+
+The CMM used MUST be the input CMM, if supplied.
+If a CMM is not supplied as the input, the decrypt operation MUST construct a [default CMM](../framework/default-cmm.md)
+from the [keyring](../framework/keyring-interface.md) inputted.
+
+The call to CMM's [Decrypt Materials](../framework/cmm-interface.md#decrypt-materials) operation
+MUST include as the input the [encryption context](../data-format/message-header.md#aad)
+(if the encryption context is empty, this operation MAY pass no encryption context),
```

Comment:
On the encrypt side we say:

- Encryption Context: If provided, this is the [input encryption context](#encryption-context).
  Otherwise, this is an empty encryption context.

I think we should say the same thing on both sides for the CMM...